src/translator/quality_estimator.h

```diff
@@ -0,0 +1,209 @@
+#pragma once
+
+#include <vector>
+
+#include "annotation.h"
+#include "response.h"
+#include "translator/history.h"
+
+namespace marian::bergamot {
+
+class QualityEstimator {
+ public:
+  /// Computes quality-scores using values from Histories and subword tokens which comes from Response
+  ///
+  ///
+  /// @param [in] histories: Histories obtained from translating a blob of source-text
+  /// @param [in] response: Partially constructed response, holding tokenization info
+  /// for source and target. The quality-scores for each sentence obtained from source-text blob
+  /// are written out as SentenceQualityEstimate into response.
+  virtual void computeQualityScores(const Histories &histories, Response &response) const = 0;
+};
+
+using SubwordRange = ByteRange;
+
+/// Unsupervised Quality Estimator model. It uses the translator model's log probabilities (log probs) as a proxy for
+/// quality scores. Then, for a given word, its quality score is computed by taking the mean of the log probs of the
+/// tokens that make it up. The sentence score is the mean of all word's log probs.
+class UnsupervisedQualityEstimator : public QualityEstimator {
+ public:
+  void computeQualityScores(const Histories &histories, Response &response) const override;
+
+ private:
+  Response::SentenceQualityEstimate computeSentenceScores(const std::vector<float> &logProbs,
+                                                          const AnnotatedText &target, const size_t sentenceIdx) const;
+};
+
+// ASCII and Unicode text files never start with the following 64 bits
+// It serves as a signature for quality estimator binary files
+constexpr std::size_t BINARY_QE_MODEL_MAGIC = 0x78cc336f1d54b180;
```

Comment:
```
/Users/runner/work/bergamot-translator/bergamot-translator/src/translator/quality_estimator.h:40:47: warning: implicit conversion from 'long long' to 'const std::size_t' (aka 'const unsigned long') changes value from 8704388732126802304 to 492089728 [-Wconstant-conversion]
constexpr std::size_t BINARY_QE_MODEL_MAGIC = 0x78cc336f1d54b180;
                      ~~~~~~~~~~~~~~~~~~~~~   ^~~~~~~~~~~~~~~~~~
```

You might want to watch out for this. /cc @abhi-agg

Reply:
Commit f8cc715